# Guided Capstone Step 4 - Pre-processing and Training Data Development

### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   **Pre-processing and Training Data Development**

5.  Modeling
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

In this step of the guided capstone, you'll revisit some of the steps you completed to solve questions in step 4. Revisiting these steps will help you get additional practice completing this kind of work.

**<font color='teal'> Start by loading the necessary packages, as you've done in the previous steps. Print out your current working directory to confirm that you are in the correct project directory. </font>**

In [1]:
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

os.getcwd()

'C:\\Users\\Billy\\Desktop\\Springboard\\Guided Capstone'

**<font color='teal'>  Load the csv file you created in step 3. Remember, it should be saved inside your data subfolder. Print the first five rows.</font>**

In [2]:
df = pd.read_csv('..\Guided Capstone\data\step3_output.csv')
df.drop('Unnamed: 0', axis='columns', inplace=True)
df.head()

Name    state  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort   Alaska         3939           2500        250      1   
1  Eaglecrest Ski Area   Alaska         2600           1540       1200      0   
2     Hilltop Ski Area   Alaska         2090            294       1796      0   
3     Arizona Snowbowl  Arizona        11500           2300       9200      0   
4  Sunrise Park Resort  Arizona        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  ...  SkiableTerrain_ac  \
0        0.0          0          2     2  ...             1610.0   
1        0.0          0          0     0  ...              640.0   
2        0.0          0          0     0  ...               30.0   
3        0.0          1          0     2  ...              777.0   
4        0.0          0          1     2  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         0  
1          53.0               90.0             0.0         0  
2          34.0              152.0            30.0         0  
3          89.0              122.0             0.0         2  
4          78.0              104.0            80.0         2  

[5 rows x 27 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `State`. Add the dummies back to the dataframe and remove the original column for `State`. </font>**

In [3]:
df_dummy = pd.concat([df.drop('state', axis=1), pd.get_dummies(df['state'])], axis=1)
df_dummy.head()

Name  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort         3939           2500        250      1   
1  Eaglecrest Ski Area         2600           1540       1200      0   
2     Hilltop Ski Area         2090            294       1796      0   
3     Arizona Snowbowl        11500           2300       9200      0   
4  Sunrise Park Resort        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  triple  ...  Rhode Island  \
0        0.0          0          2     2       0  ...             0   
1        0.0          0          0     0       0  ...             0   
2        0.0          0          0     0       1  ...             0   
3        0.0          1          0     2       2  ...             0   
4        0.0          0          1     2       3  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 61 columns]

In [4]:
df_dummy.shape

(330, 61)

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing, standardize the scale of the features of the dataframe except the name of the resort, which you don't need in the dataframe for modeling so it can be droppped here as well. You should  hold out your response variable(s) so that you can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as your response for scaling and modeling. Later, you will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave, them in the development dataframe. </font>**

In [5]:
y = df_dummy['AdultWeekend'].ravel()
X = df_dummy.drop(['Name', 'AdultWeekend'], axis=1)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selectionm import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. You will start by using the adult weekend ticket price as your response variable for modeling.</font>**

In [6]:
from sklearn.model_selection import train_test_split
from sklearn import preprocessing

In [7]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=17)

In [8]:
scaler = preprocessing.StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

Here, you'll start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5 - Modeling

You've cleaned and prepared the datasets, so now it's time to get into the most exciting part of this process: modeling! In this exercise, you will build three different models and compare each model's performance. In the end, you will choose the best model for proving insights to Big Mountain management.  

1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with Training Data Set

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [9]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score, mean_absolute_error

lm = linear_model.LinearRegression()
model = lm.fit(X_train, y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

In [10]:
y_pred = model.predict(X_test)

In [11]:
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

-2.24333771587819e+22
462353067640.7014


## Review Model Outcomes — Iterate over additional models as needed.

**<font color='teal'> Print the intercept value from the linear model. </font>**

In [12]:
lm.intercept_

63.332835786337185

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the `AdultWeekend` resulting value. Also, because you took the time to scale your x values in the training data, you can compare each of the coefficients for the features to determine the feature's importance. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 

Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what you are looking for is the magnitude of impact on your response variable.

In [13]:
coeff = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])  
coeff.sort_values(by=['Coefficient'],ascending=False).head(10)

Coefficient
fastEight     1.260569e+13
total_chairs  8.685166e+12
New York      7.786692e+12
Colorado      7.005501e+12
Michigan      6.833591e+12
California    6.655541e+12
Pennsylvania  6.079067e+12
Utah          5.423096e+12
Minnesota     5.423096e+12
Washington    5.181663e+12

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that you care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

#### Model 2

In [14]:
#select original df and drop Name, AdultWeekend, and state columns
X = df.drop(['Name', 'AdultWeekend', 'state'], 1)
X.head()

summit_elev  vertical_drop  base_elev  trams  fastEight  fastSixes  \
0         3939           2500        250      1        0.0          0   
1         2600           1540       1200      0        0.0          0   
2         2090            294       1796      0        0.0          0   
3        11500           2300       9200      0        0.0          1   
4        11100           1800       9200      0        0.0          0   

   fastQuads  quad  triple  double  ...  LongestRun_mi  SkiableTerrain_ac  \
0          2     2       0       0  ...            1.0             1610.0   
1          0     0       0       4  ...            2.0              640.0   
2          0     0       1       0  ...            1.0               30.0   
3          0     2       2       1  ...            2.0              777.0   
4          1     2       3       1  ...            1.2              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   projectedDaysOpen  NightSkiing_ac  clusters  
0              150.0           550.0         0  
1               90.0             0.0         0  
2              152.0            30.0         0  
3              122.0             0.0         2  
4              104.0            80.0         2  

[5 rows x 24 columns]

In [15]:
#split into training and testing datasets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=17)

In [16]:
#standardize the scale of the features
scaler = preprocessing.StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [17]:
#train model
lm = linear_model.LinearRegression()
model = lm.fit(X_train, y_train)

In [18]:
#predict y
y_pred = model.predict(X_test)

In [19]:
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

0.8387988040009055
6.716305811316018


In [20]:
#intercept value of the model
lm.intercept_

63.33288421633073

In [21]:
coeff = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])  
coeff.sort_values(by=['Coefficient'],ascending=False).head(10)

Coefficient
AdultWeekday        20.623286
base_elev            4.525014
summit_elev          3.726818
quad                 1.307021
averageSnowfall      1.301179
NightSkiing_ac       1.222188
triple               1.053043
daysOpenLastYear     0.928379
surface              0.834357
Snow Making_ac       0.667220

**<font color='teal'> When reviewing your new model coefficients, note that `summit_elev` is now in the number two spot. From a managerial perspective, this is also difficult to change and highly correlated with `base_elev` and `vertical_drop`.  This time rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [22]:
#select original df and drop Name, AdultWeekend, state, and summit_elev columns
X = df.drop(['Name', 'AdultWeekend', 'state', 'summit_elev', 'base_elev'], 1)
X.head()

vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  triple  \
0           2500      1        0.0          0          2     2       0   
1           1540      0        0.0          0          0     0       0   
2            294      0        0.0          0          0     0       1   
3           2300      0        0.0          1          0     2       2   
4           1800      0        0.0          0          1     2       3   

   double  surface  total_chairs  ...  LongestRun_mi  SkiableTerrain_ac  \
0       0        2             7  ...            1.0             1610.0   
1       4        0             4  ...            2.0              640.0   
2       0        2             3  ...            1.0               30.0   
3       1        2             8  ...            2.0              777.0   
4       1        0             7  ...            1.2              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   projectedDaysOpen  NightSkiing_ac  clusters  
0              150.0           550.0         0  
1               90.0             0.0         0  
2              152.0            30.0         0  
3              122.0             0.0         2  
4              104.0            80.0         2  

[5 rows x 22 columns]

In [23]:
#split into training and testing datasets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=17)

In [24]:
#standardize the scales of the features
scaler = preprocessing.StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [25]:
#train model
model = lm.fit(X_train, y_train)

In [26]:
#predict y
y_pred = model.predict(X_test)

In [27]:
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

0.8393725398169642
6.68880829014327


In [28]:
#model intercept
lm.intercept_

63.33288421633073

In [29]:
coeff = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])  
coeff.sort_values(by=['Coefficient'],ascending=False).head(10)

Coefficient
AdultWeekday        20.627397
vertical_drop        1.508742
averageSnowfall      1.445002
quad                 1.344216
NightSkiing_ac       1.297494
clusters             1.037104
triple               1.018248
surface              0.887830
daysOpenLastYear     0.802297
fastQuads            0.714952

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell. You will explain your selection during the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -2.24333771587819e+22 | 462353067640.7014 |-|
| Model 2. | 0.84 | 6.72 |'state'|
| Model 3. | 0.84 | 6.69 |'state','summit_elev','base_elev'|

Model Selection: 3